rfd/0058-package-distribution.md

```diff
@@ -0,0 +1,66 @@
+| Notes:                                     |                            | Can do anything we want, just depends on the amount of initial and reoccuring engineering effort is required | Higher complexity, but supports pretty much any use case we'd ever need | Easy to use but probably not the best solution | Still in preview, not generally available                                         |
+
+### Signing key management
+JFrog Artifactory and PackageCloud require handing over our repo signing keys to them, which is a non-starter. While GCP requires using their own signing key (which is used for all GCP repositories in a given region), it is assumed that their security is sufficient to protect said key. Lastly, fixing the current solution will keep us in control of the key, but will require us to continuing storing and securing it ourselves.
+
+### Channel support
+#### APT
+The current solution does have support for APT. GCP Artifact Registry supports APT channels only via the "distribution" parameter of an APT source. This is also directly tied to the registry name inside of GCP. In addition, the URL for the repository depends on the GCP project ID which could cause issues for all clients if the GCP project was recreated or we wanted to change which project owned the repository.
+
+Fixing the current solution would allow for channel support via standard practices for APT using APT distribution and component parameters. This is inline with what most APT repositories do, and supports the scheme that teleport users are most familiar with.
+
+#### YUM
+As with APT, the current solutoin does not support YUM channels. GCP would add support, but only by means of the project ID and repository name. As with APT this is a somewhat fragile and inflexible solution. Fixing the current solution would allow for channel support in whatever manner we like, allowing it to conform to best practices and standard naming scheme.
+
+### Recommendations
+It is worth noting that GCP Artifact Registry is in preview and has not had an update since 11/2021. It's lacking some features as outlined above as well as some other we may wish to use in the future (i.e. bringing our own key). In addition, if there was a disaster recovery event where we needed to push to an Artifact Registry in another region while GCP had an outage, we would break `apt update` and `yum update` for our customers as the key they would use is region-specific.
+
+As discussed above, JFrog Artifactory and PackageCloud are non-starters due to their signing key requirements.
+
+As a consequence of the the issues with the third party solutions, along with the additional channel support features available, I recommend fixing our current S3-hosted solution. This would require switching reprepo for aptly in our Drone pipelines, and updating the current createrepo steps.
```

Comment:
I'd like a bit more detail on the changes.  E.g.
What aptly commands will generate the right metadata?
How will RPM publishing need to change, if at all?
Are there any concurrency concerns?
How will pre-releases be handed (if at all)?

Can we add old releases (e.g. backpopulate an earlier patch release) without affecting the latest version?

Reply:
Added some info here: https://github.com/gravitational/teleport/blob/rfd/0058-package-distribution/rfd/0058-package-distribution.md#research. There could be some concurrency issues, but no more than what already exists.